src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.iOS.xml

```diff
@@ -3,5 +3,8 @@
     <type fullname="System.Runtime.CompilerServices.RuntimeFeature">
       <method signature="System.Boolean get_IsDynamicCodeCompiled()" body="stub" value="false" />
     </type>
+    <type fullname="System.Runtime.CompilerServices.RuntimeFeature" feature="System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported" featurevalue="false">
```

Comment:
What is the point of this over the existing:

https://github.com/dotnet/runtime/blob/35996db345d0d0645cb5df14debacbadb88cf9b2/src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml#L7-L10

Isn't this duplicated with the shared logic?

Reply:
`IsDynamicCodeCompiled` and `IsDynamicCodeSupported` are different things that cannot be grouped under the same feature flag (eg. IL interpreter is `false`/`true`).

Reply:
I understand that they are different flags, and I understand the difference between them.

What I'm asking is "why are we making this change specific to iOS when the same substitution XML exists for all platforms?"

Reply:
As an aside: looking at the full existing logic in the shared substitutions file, there is probably a bug with it:

https://github.com/dotnet/runtime/blob/35996db345d0d0645cb5df14debacbadb88cf9b2/src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml#L3-L10

Line 4 should be deleted. Just because Dynamic Code "is supported" is true, doesn't mean that Dynamic Code "is compiled" should be true as well. For `false`, the existing logic is fine - if you can't support Dynamic Code, then it follows that it can't be compiled.

Reply:
Sorry, I misunderstood the question given how GitHub mobile web framed the code listing. I see what you mean now, and I agree that the shared substitutions don't work for the interpreter case.

Reply:
> What is the point of this over the existing:
> 
> https://github.com/dotnet/runtime/blob/35996db345d0d0645cb5df14debacbadb88cf9b2/src/coreclr/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.xml#L7-L10
> 
> Isn't this duplicated with the shared logic?

If I am not mistaken, the substitution file constructed when building `System.Private.CoreLib.dll` for iOS platforms does not seem to include the substitution file you linked.

The `ILLink.Substitutions.xml` for iOS `System.Private.CoreLib.dll` is constructed from the following pieces:
- according to: https://github.com/dotnet/runtime/blob/36439c510b779103a4a8066359d0d63dc003eed3/src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj#L155
    - includes:
    1. https://github.com/dotnet/runtime/blob/main/src/mono/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.iOS.xml

- according to: https://github.com/dotnet/runtime/blob/36439c510b779103a4a8066359d0d63dc003eed3/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems#L53-L61
    - includes:

    1. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml
    2. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.LittleEndian.xml
    3. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.64bit.xml
    4. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.NoX86Intrinsics.xml
    5. https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.OSX.xml

- according to: https://github.com/dotnet/runtime/blob/36439c510b779103a4a8066359d0d63dc003eed3/eng/illink.targets#L124
    - includes: 

    1. https://github.com/dotnet/runtime/blob/main/eng/ILLink.Substitutions.Resources.template
    
Resulting in the end with: https://gist.github.com/ivanpovazan/6a97df9974a7f36b596837edd26b87ec
NOTE: The gist above shows `ILLink.Substitutions.xml` from a disassembled `System.Private.CoreLib.dll` taken from `microsoft.netcore.app.runtime.ios-arm64` 

---
It might be that the build process for iOS-like platforms is referencing the wrong shared substitution file, but according to the current set up I changed the one which is included in the build. Should this be changed?

Regarding your suggestion in: https://github.com/dotnet/runtime/pull/86971#discussion_r1231473969 I can make the required changes, but we should agree/clarify on which substitution file gets included in this case :)

Reply:
You are right - I was mistaken above. I misread the existing code.

The `ILLink.Substitutions.xml` file I pointed to above is specific for "coreclr" - it is under `src/coreclr/System.Private.CoreLib/...`. I was thinking this was the "ILLink.Substitutions.Shared.xml" file under `src/libraries`, which is used for all platforms. Given this, both concerns I had above are invalid. The existing logic is right, and the change you are making is necessary.